### DISCUSSION TITLE
SOS 🚨​: Training ASR on Arabic/English data predicts blank token after ~100 steps and model is not learning.

### DISCUSSION THREAD

**Discussion by Houss3m** [NVIDIA-NeMo/NeMo#14614] [Q&A]
Hello team;

I am trying to train an ASR model using NeMo framework, tried many recipies, but all of them are predicting blank for my dataset.

details:
- my dataset has mix of Arabic and English, transcriptions for Arabic are written in Arabic script, while the English are written in Latin scripts.
- When I do train a conformer Tranducer for example, it starts predicting junk transcriptions, after a number of steps, it starts predicting blanks (nothing).
- like the model is not learning and see the blank is the best path and keeps predicting blank only.
- I doubted the dataset, so I created two splits, one for English and one for Arabic.
- I trained on the English data, the model started predicting junk text, then it started to predict blank, then it started to learn after 40% of the first epoch and performed well now.
- I did the same steps for the Arabic split (building tokenizer, tarred the data etc..), and used the same config and parameters but the model starts predicting blank after certain number of steps and it is stuck there.
- the tokenizer works well when I test it on text data.
- I increased the vocab size to 4k and still same issue... 
- tried to reduce learning rate and still the same issue.

any thoughts on what could be the issue? like encoding of the json file can be an issue? 

example of the outputs:

```txt
[NeMo I wer:330] reference:مسبح
[NeMo I wer:331] predicted:ائرائرائرائرائرائرائرائرائرائرلاداهلادلادوةوةنيلي يكنيلي يكن يكن يكن اي يكن يكن يكن يكن يكن 
[NeMo I wer:329] 
    
[NeMo I wer:330] reference:اربع
[NeMo I wer:331] predicted:حيحيحيحيحيحيحيؤولحيحي ده فيده ده ده ده ده ده ده ده دصيرتيصير اف اف ا
.....

# after many steps I get blank and keeps going to epoch 10:

[NeMo I wer:330] reference:وليس هزيمتهم بالامر الذي نتوقعه
[NeMo I wer:331] predicted:

Epoch 7:  37%|███▋      | 10114/27008 [55:21<1:32:27,  3.05it/s, v_num=0, train_step_timing in s=0.362]
```

samples of the date:
```
{"audio_filepath": "path0.flac", "duration": 1.17, "text": "تشربو شي حاجة", "shard_id": 0}
{"audio_filepath": "path1.flac", "duration": 1.31, "text": "خمس", "shard_id": 0}
{"audio_filepath": "path2.flac", "duration": 1.32, "text": "شنو واقع", "shard_id": 0}
```

I build my tokenizer using this command:
```
python process_asr_text_tokenizer.py \
  --manifest=valid_arabic.jsonl,train_arabic.jsonl \
  --vocab_size=4096 \
  --data_root=arabonly_4096 \
  --tokenizer="spe" \
  --spe_type=bpe \
  --spe_character_coverage=1.0 \
  --spe_max_sentencepiece_length=16 \
  --log
```

any insights please?

**Comment by nithinraok** [comment]
Hi, 

check tokenizer encode and decode options to validate for arabic you are able to see the text as intended.  

Start with transducer instead of hybrid and also start from a pretrained encoder (like parakeet-tdt-0.6b-v3). Would also recommend to use fastconformer rather than conformer configs.

**Reply by Houss3m** [reply]
Hi Nithinraok, thanks for your time,

I checked the tokenizer and validate that it works well for my data in Arabic script as it should, I tried to save the json manifest file with ensure_ascii=True. but still getting blanks and the model is not learning. 

see below for the tokenizer:
**Text:** ومن فاس محمد فقيهي استاذ القانون الدستوري والعلوم السياسية في جامعة محمد بن عبد الله نبدا اولا مع السيد توفيق بوعشرين سيد توفيق هذه الانتخابات المحلية في المغرب الذي يقال بانه استطاع ان يعبر الربيع العربي باقل الخسائر بفضل ح

**Tokens:** ['▁ومن', '▁ف', 'اس', '▁محمد', '▁ف', 'قي', 'هي', '▁استاذ', '▁القانون', '▁الدستوري', '▁وال', 'عل', 'وم', '▁السياسية', '▁في', '▁جامعة', '▁محمد', '▁بن', '▁عبد', '▁الله', '▁نبدا', '▁اولا', '▁مع', '▁السيد', '▁تو', 'فيق', '▁ب', 'وع', 'شرين', '▁سيد', '▁تو', 'فيق', '▁هذه', '▁الانتخابات', '▁المحلية', '▁في', '▁المغرب', '▁الذي', '▁يقال', '▁بانه', '▁استطاع', '▁ان', '▁يعبر', '▁الربيع', '▁العربي', '▁با', 'قل', '▁الخ', 'س', 'ائر', '▁ب', 'فضل', '▁ح']

**IDs:** [376, 7, 186, 611, 7, 105, 580, 823, 840, 2826, 48, 116, 67, 471, 15, 2058, 611, 478, 443, 252, 3209, 676, 65, 576, 1553, 2873, 6, 154, 1988, 318, 1553, 2873, 73, 1053, 4019, 15, 2126, 160, 2468, 961, 4123, 16, 4637, 2339, 584, 241, 215, 153, 4968, 899, 6, 781, 28]

**Reconstructed:** ومن فاس محمد فقيهي استاذ القانون الدستوري والعلوم السياسية في جامعة محمد بن عبد الله نبدا اولا مع السيد توفيق بوعشرين سيد توفيق هذه الانتخابات المحلية في المغرب الذي يقال بانه استطاع ان يعبر الربيع العربي باقل الخسائر بفضل ح


my code:
```
import sentencepiece as spm

# Load model
sp = spm.SentencePieceProcessor()
sp.load("tokenizer.model")

# Test
text = "\u0648\u0645\u0646 \u0641\u0627\u0633 \u0645\u062d\u0645\u062f \u0641\u0642\u064a\u0647\u064a \u0627\u0633\u062a\u0627\u0630 \u0627\u0644\u0642\u0627\u0646\u0648\u0646 \u0627\u0644\u062f\u0633\u062a\u0648\u0631\u064a \u0648\u0627\u0644\u0639\u0644\u0648\u0645 \u0627\u0644\u0633\u064a\u0627\u0633\u064a\u0629 \u0641\u064a \u062c\u0627\u0645\u0639\u0629 \u0645\u062d\u0645\u062f \u0628\u0646 \u0639\u0628\u062f \u0627\u0644\u0644\u0647 \u0646\u0628\u062f\u0627 \u0627\u0648\u0644\u0627 \u0645\u0639 \u0627\u0644\u0633\u064a\u062f \u062a\u0648\u0641\u064a\u0642 \u0628\u0648\u0639\u0634\u0631\u064a\u0646 \u0633\u064a\u062f \u062a\u0648\u0641\u064a\u0642 \u0647\u0630\u0647 \u0627\u0644\u0627\u0646\u062a\u062e\u0627\u0628\u0627\u062a \u0627\u0644\u0645\u062d\u0644\u064a\u0629 \u0641\u064a \u0627\u0644\u0645\u063a\u0631\u0628 \u0627\u0644\u0630\u064a \u064a\u0642\u0627\u0644 \u0628\u0627\u0646\u0647 \u0627\u0633\u062a\u0637\u0627\u0639 \u0627\u0646 \u064a\u0639\u0628\u0631 \u0627\u0644\u0631\u0628\u064a\u0639 \u0627\u0644\u0639\u0631\u0628\u064a \u0628\u0627\u0642\u0644 \u0627\u0644\u062e\u0633\u0627\u0626\u0631 \u0628\u0641\u0636\u0644 \u062d"
tokens = sp.encode(text, out_type=str)

ids = sp.encode(text, out_type=int)
reconstructed = sp.decode(ids)

print("Text:", text)
print("Tokens:", tokens)
print("IDs:", ids)
print("Reconstructed:", reconstructed)
```


I tried using fastconformer as you advised, using the following python and config files, but still getting blank in predictions, how many epochs you think it would take for the model to learn?

Note: I am using longform dataset that's why I am experimenting with this config.

code: 
```
python ./nemo/NeMo/examples/asr/asr_ctc/speech_to_text_ctc_bpe.py \
    --config-path=./nemo/NeMo/examples/asr/conf/fastconformer/long_fastconformer/ \
    --config-name=fast-conformer-long_ctc_bpe \
    \
    exp_manager.name="fastconformernomdaspc" \
    exp_manager.resume_if_exists=true \
    exp_manager.resume_ignore_no_checkpoint=true \
    exp_manager.exp_dir=./nemo/NeMo/experiments/fastconformer \
    model.tokenizer.dir=$TOKENIZER \
    model.tokenizer.type=bpe \
    model.train_ds.manifest_filepath=$TRAIN_MANIFEST \
    model.validation_ds.manifest_filepath=$VAL_MANIFEST \
    trainer.devices=1 \
    trainer.accelerator=gpu
```



I tried running the training with an encoder model "stt_en_conformer_transducer_small", I did freeze to the encoder and trained the decoder, the validation loss didn't improve much and went back to higher values. 
I used the following params:
```
optimizer_conf = {}

optimizer_conf['name'] = 'adamw'
optimizer_conf['lr'] = 0.001
optimizer_conf['betas'] =  [0.9, 0.98]
optimizer_conf['weight_decay'] = 0

sched = {}
sched['name'] = 'CosineAnnealing'
sched['warmup_steps'] = None
sched['warmup_ratio'] = 0.10
sched['min_lr'] = 1e-6
optimizer_conf['sched'] = sched
```


<img width="516" height="868" alt="image" src="https://github.com/user-attachments/assets/07f1b19d-b895-4b75-978d-31ef95d94d41" />


<img width="422" height="868" alt="image" src="https://github.com/user-attachments/assets/da2df469-c04a-431e-8d7d-8629571b8864" />

**Reply by mingewang** [reply]
@Houss3m, you mentioned that "it started to learn after 40% of the first epoch and performed well afterward."
I’m curious — how many hours of English data are in your dataset? I’m using the hybrid FastConformer config and facing a similar issue: the model outputs blank predictions after many 150 epochs  (with around 3,000 hours of English data).

**Reply by Houss3m** [reply]
@mingewang I was using the librispeech dataset (960h) for the english experiment

**Comment by nithinraok** [comment]
Since you are using 1 GPU, try these:
increase grad_accumulation to like 4 or 8 
reduce lr 
increase warmup 0.2

add data if possible. 

Try finetuning rather than training from scratch

**Comment by gnomefin** [comment]
@nithinraok your suggestion works for 1 GPU, lowering the lr, in my case I am training ctc+rnnt hybrid from en base model, and build arabic tokenizer from scratch.

```

python3 "$(basename "$TRAIN_SCRIPT")" \
    --config-path=conf/fastconformer/hybrid_transducer_ctc \
    --config-name=fastconformer_hybrid_transducer_ctc_bpe \
    exp_manager.name="arabic_fastconformer" \
    exp_manager.resume_if_exists=true \
    exp_manager.resume_ignore_no_checkpoint=false \
    +exp_manager.checkpoint_callback_params.save_last=true \
    exp_manager.exp_dir="$ABS_EXP_DIR" \
    model.tokenizer.dir="$ABS_TOKENIZER" \
    model.train_ds.manifest_filepath="$ABS_TRAIN_MANIFEST" \
    model.train_ds.num_workers=8 \
    model.validation_ds.manifest_filepath="$ABS_VAL_MANIFEST" \
    model.validation_ds.num_workers=8 \
    model.test_ds.manifest_filepath="$ABS_TEST_MANIFEST" \
    model.test_ds.num_workers=8 \
    model.loss.loss_name="pytorch" \
    model.aux_ctc.ctc_loss_weight=0.3 \
    model.aux_ctc.decoding.strategy="greedy_batch" \
    $TARRED_ARGS \
    trainer.devices=1 \
    trainer.accelerator=gpu \
    trainer.max_epochs=30 \
    trainer.accumulate_grad_batches=8 \
    model.optim.name=adamw \
    model.optim.lr=1e-4 \
    model.optim.sched.warmup_steps=null \
    model.optim.sched.warmup_ratio=0.2
```

1e-3 blank result,
5e-4 still blank, doesn't work,

1e-4 better.

however in the second epochs, it starts to blank sometimes:

```

[NeMo I 2025-12-02 09:18:24 nemo_logging:393] WER reference:ثم جاءهم ما كانوا يوعدون
[NeMo I 2025-12-02 09:18:24 nemo_logging:393] WER predicted:
Epoch 2:  18%|████████████▌                                                          | 318/1792 [11:29<53:16,  0.46it/s, v_num=4, train_step_timing in s=1.260][NeMo I 2025-12-02 09:18:25 nemo_logging:393]

[NeMo I 2025-12-02 09:18:25 nemo_logging:393] WER reference:وظلم فغفر
[NeMo I 2025-12-02 09:18:25 nemo_logging:393] WER predicted:ثهثهثهثهثهثهثهثهثهثه خطلهوق
[NeMo I 2025-12-02 09:18:25 nemo_logging:393]

[NeMo I 2025-12-02 09:18:25 nemo_logging:393] WER reference:فاذا النجوم طمست
[NeMo I 2025-12-02 09:18:25 nemo_logging:393] WER predicted:
[NeMo I 2025-12-02 09:18:25 nemo_logging:393]

[NeMo I 2025-12-02 09:18:25 nemo_logging:393] WER reference:قل لله الشفاعه جميعا له ملك السماوات والارض ثم اليه ترجعون
[NeMo I 2025-12-02 09:18:25 nemo_logging:393] WER predicted:تاج ا ا ا ا ا ا ا ا ا فما فما فما فما فما خط خط خط خط خط خط خط خط خط خط
[NeMo I 2025-12-02 09:18:26 nemo_logging:393]

[NeMo I 2025-12-02 09:18:26 nemo_logging:393] WER reference:فانتقمنا منهم وانهما لبامام مبين
[NeMo I 2025-12-02 09:18:26 nemo_logging:393] WER predicted:بيخرج بيخرجخرج بيخرجخرج بيخرج
[NeMo I 2025-12-02 09:18:26 nemo_logging:393]

[NeMo I 2025-12-02 09:18:26 nemo_logging:393] WER reference:وظلم فغفر
[NeMo I 2025-12-02 09:18:26 nemo_logging:393] WER predicted:
Epoch 2:  18%|████████████▋                                                          | 319/1792 [11:31<53:15,  0.46it/s, v_num=4, train_step_timing in s=2.270][NeMo I 2025-12-02 09:18:27 nemo_logging:393]

[NeMo I 2025-12-02 09:18:27 nemo_logging:393] WER reference:لماذا تتاخر دائما
[NeMo I 2025-12-02 09:18:27 nemo_logging:393] WER predicted:الد الد الد الد الد الد الد الد الد الد خط
[NeMo I 2025-12-02 09:18:27 nemo_logging:393]

[NeMo I 2025-12-02 09:18:27 nemo_logging:393] WER reference:عاده ما ينام بعد العشاء
[NeMo I 2025-12-02 09:18:27 nemo_logging:393] WER predicted:خط خط خط خط خط خط خط خط خط خط
[NeMo I 2025-12-02 09:18:28 nemo_logging:393]

[NeMo I 2025-12-02 09:18:28 nemo_logging:393] WER reference:وان لو استقاموا علي الطريقه لاسقيناهم ماء غدقا
[NeMo I 2025-12-02 09:18:28 nemo_logging:393] WER predicted:خط خط خط خط خط خط خط خط خط خط
[NeMo I 2025-12-02 09:18:28 nemo_logging:393]

[NeMo I 2025-12-02 09:18:28 nemo_logging:393] WER reference:قالوا بل انتم لا مرحبا بكم انتم قدمتموه لنا فبئس القرار
[NeMo I 2025-12-02 09:18:28 nemo_logging:393] WER predicted:
[NeMo I 2025-12-02 09:18:28 nemo_logging:393]

[NeMo I 2025-12-02 09:18:28 nemo_logging:393] WER reference:لماذا تتاخر دائما
[NeMo I 2025-12-02 09:18:28 nemo_logging:393] WER predicted:
Epoch 2:  18%|████████████▉                                                          | 327/1792 [11:47<52:47,  0.46it/s, v_num=4, train_step_timing in s=1.690]
```

I am currently still waiting until the last epochs.

**Reply by kavinask007** [reply]
any update on this ? I'm facing the same issue for fastconformer_transducer_bpe_streaming_prompt. Model goes into blank collapse.
